### ISSUE TITLE
Error: "Problem instantiating module class from EXPath repository: org.exist.xquery.modules.expath.file.ExpathFileModule"

### ISSUE THREAD

**Issue by joewiz** [eXist-db/expath-file-module#4]
When I install the package, quit eXist, comment out the native file module in conf.xml, restart eXist (develop c018bd44b from today), and run a sample query importing the EXPath file module, I get an error in exist.log.

**sample query:**

```xquery
xquery version "3.1";

import module namespace f="http://expath.org/ns/file";

"test"
```

**exist.log:**

```text
2017-06-28 19:40:51,675 [qtp94682872-21] ERROR (XQueryServlet.java [process]:534) - Cannot compile xquery: exerr:ERROR error found while loading module f: Problem instantiating module class from EXPath repository: org.exist.xquery.modules.expath.file.ExpathFileModule 
org.exist.EXistException: Cannot compile xquery: exerr:ERROR error found while loading module f: Problem instantiating module class from EXPath repository: org.exist.xquery.modules.expath.file.ExpathFileModule
	at org.exist.http.servlets.XQueryServlet.process(XQueryServlet.java:428) [exist-optional.jar:3.3.0-SNAPSHOT]
	at org.exist.http.servlets.XQueryServlet.doPost(XQueryServlet.java:191) [exist-optional.jar:3.3.0-SNAPSHOT]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707) [servlet-api-3.1.jar:3.1.0]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) [servlet-api-3.1.jar:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:845) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:583) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:566) [jetty-security-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1174) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1106) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:159) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:74) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.exist.http.urlrewrite.Forward.doRewrite(Forward.java:50) [exist-optional.jar:3.3.0-SNAPSHOT]
	at org.exist.http.urlrewrite.XQueryURLRewrite.doRewrite(XQueryURLRewrite.java:545) [exist-optional.jar:3.3.0-SNAPSHOT]
	at org.exist.http.urlrewrite.XQueryURLRewrite.service(XQueryURLRewrite.java:349) [exist-optional.jar:3.3.0-SNAPSHOT]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) [servlet-api-3.1.jar:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:845) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1689) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at de.betterform.agent.web.filter.XFormsFilter.doFilter(XFormsFilter.java:171) [betterform-exist-5.1-SNAPSHOT-20160615.jar:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:581) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524) [jetty-security-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1174) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1106) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:213) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:119) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:396) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.Server.handle(Server.java:524) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:319) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:253) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
Caused by: org.exist.xquery.XPathException: exerr:ERROR error found while loading module f: Problem instantiating module class from EXPath repository: org.exist.xquery.modules.expath.file.ExpathFileModule
	at org.exist.repo.ExistRepository.getModule(ExistRepository.java:135) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.repo.ExistRepository.resolveJavaModule(ExistRepository.java:110) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQueryContext.resolveInEXPathRepository(XQueryContext.java:346) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQueryContext.importModule(XQueryContext.java:2690) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.importDecl(XQueryTreeParser.java:5862) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.prolog(XQueryTreeParser.java:4877) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.mainModule(XQueryTreeParser.java:4010) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.module(XQueryTreeParser.java:3955) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.xpath(XQueryTreeParser.java:3612) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQuery.compile(XQuery.java:128) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQuery.compile(XQuery.java:79) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQuery.compile(XQuery.java:71) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.http.servlets.XQueryServlet.process(XQueryServlet.java:425) ~[exist-optional.jar:3.3.0-SNAPSHOT]
	... 48 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_131]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_131]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_131]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_131]
	at org.exist.repo.ExistRepository.instantiateModule(ExistRepository.java:151) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.repo.ExistRepository.getModule(ExistRepository.java:124) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.repo.ExistRepository.resolveJavaModule(ExistRepository.java:110) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQueryContext.resolveInEXPathRepository(XQueryContext.java:346) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQueryContext.importModule(XQueryContext.java:2690) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.importDecl(XQueryTreeParser.java:5862) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.prolog(XQueryTreeParser.java:4877) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.mainModule(XQueryTreeParser.java:4010) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.module(XQueryTreeParser.java:3955) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.xpath(XQueryTreeParser.java:3612) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQuery.compile(XQuery.java:128) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQuery.compile(XQuery.java:79) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQuery.compile(XQuery.java:71) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.http.servlets.XQueryServlet.process(XQueryServlet.java:425) ~[exist-optional.jar:3.3.0-SNAPSHOT]
	... 48 more
Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.exist.xquery.modules.expath.file.ExpathFileModule$
	at org.exist.xquery.modules.expath.file.ExpathFileModule.<init>(ExpathFileModule.scala:77) ~[expath-file-module-1.0-SNAPSHOT.jar:?]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_131]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_131]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_131]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_131]
	at org.exist.repo.ExistRepository.instantiateModule(ExistRepository.java:151) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.repo.ExistRepository.getModule(ExistRepository.java:124) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.repo.ExistRepository.resolveJavaModule(ExistRepository.java:110) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQueryContext.resolveInEXPathRepository(XQueryContext.java:346) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQueryContext.importModule(XQueryContext.java:2690) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.importDecl(XQueryTreeParser.java:5862) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.prolog(XQueryTreeParser.java:4877) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.mainModule(XQueryTreeParser.java:4010) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.module(XQueryTreeParser.java:3955) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.parser.XQueryTreeParser.xpath(XQueryTreeParser.java:3612) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQuery.compile(XQuery.java:128) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQuery.compile(XQuery.java:79) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.xquery.XQuery.compile(XQuery.java:71) ~[exist.jar:3.3.0-SNAPSHOT]
	at org.exist.http.servlets.XQueryServlet.process(XQueryServlet.java:425) ~[exist-optional.jar:3.3.0-SNAPSHOT]
	... 48 more
```

**Comment by adamretter** [comment]
Hmm... there seems to be some sort of bug in the expath-repo in eXist-db. I spent some time trying to debug it today, but I am struggling to understand how it should work. Suffice to say it isn't correctly loading the XAR file.

As a workaround, you can do the following:

1. Copy the file `target/expath-file-module-1.0-SNAPSHOT-exist.jar` file into `$EXIST_HOME/lib/user`

2. Replace the line in `$EXIST_HOME/conf.xml`:
```xml
<module uri="http://exist-db.org/xquery/file" class="org.exist.xquery.modules.file.FileModule" />
```
with
```
 <module uri="http://expath.org/ns/file" class="org.exist.xquery.modules.expath.file.ExpathFileModule"/>
```

3. Restart eXist-db.

4. Run the query (MacOS assumed):
```xquery
declare namespace file="http://expath.org/ns/file";
file:exists("/Users")
```

**Comment by adamretter** [comment]
@joewiz I think this should now be fixed. Can you test again, pulling in the latest source code?

**Comment by joewiz** [comment]
By the way, my immediate interest in digging back into this was hoping that I might begin to benefit from the better error messages defined in the EXPath File Spec. For example, eXist's native file had me very confused with this message:

> exerr:ERROR Cannot serialize file. A problem occurred while serializing the binary data: /Users/joe/Downloads/test/test.txt [at line 9, column 5]

The test query that was failing:

```xquery
xquery version "3.1";

let $dir := "/Users/joe/Downloads/test"
let $filename := "test.txt"
let $path := $dir || "/" || $filename
let $string := "foo"
let $binary := util:string-to-binary($string)
return
    file:serialize-binary($binary, $path)
```

The actual source of the error was that the "test" directory didn't exist. In contrast, the EXPath File Module Spec defines an error code for precisely this scenario:

> Error Conditions
> 
> `file:no-dir` is raised if the parent directory of `$file` does not exist.
> ...

(See http://expath.org/spec/file#fn.write.)

I considered filing an issue on the main eXist repository, but if this error handling is already implemented here, I'd most likely switch my code over to this module instead.

**Comment by joewiz** [comment]
I notice that updates to this issue aren't being registered on the HipChat eXistdb "integrations" room. I looked at some of the repositories where it is configured, and couldn't see a HipChat integration registered there.

**Comment by adamretter** [comment]
@joewiz So first thing first. The error you are getting with the build seems to be specific to Maven on your machine. A couple things to try (in no order):

* Can you try updating to the latest version of Apache Maven and see if that fixes it?
* Deleting the cached version of the `maven-jar-plugin` which you seem to be having problems with: `rm -rf ~/.m2/repository/org/apache/maven/plugins/maven-jar-plugin`

If that doesn't help, can you run `mvn` with the `-X` flag which should give us plenty of debugging info...

**Comment by adamretter** [comment]
@joewiz Yes the EXPath file module should have better error codes and responses.

**Comment by adamretter** [comment]
@joewiz I am not quite sure what you mean about HipChat?

**Comment by joewiz** [comment]
@adamretter Thanks! Based on https://maven.apache.org/download.cgi, the 3.5.2 version of Maven that I am running is indeed the latest version. I ran your command to delete `maven-jar-plugin`, but re-running failed with the same error. Next, I tried just blowing away the entire `~/.m2` directory and running the README's build directions again for both repositories. This time it worked. I'm not sure if there is a word in maven land for a corrupt local repository, but I think that must've been the cause here, and I think it has happened before. I wish I had better tools for self-diagnosing, but thank you for helping point me in the right direction! Next up, I'll try installing, using, etc.
